webui/src/pages/auth/login.tsx

```diff
@@ -27,11 +29,48 @@ export interface LoginConfig {
     logout_url: string;
 }
 
-const LoginForm = ({loginConfig}: {loginConfig: LoginConfig}) => {
+type NavigateState = { redirected?: boolean; next?: string };
+
+const withNext = (url: string, next: string) => {
+    const safeNext = next?.startsWith("/") ? next : "/";
+    try {
+        const u = new URL(url, window.location.origin);
+        u.searchParams.set("next", safeNext);
+        return u.toString();
+    } catch {
+        return `${url}${url.includes("?") ? "&" : "?"}next=${encodeURIComponent(safeNext)}`;
```

Comment:
This is both hard to read and tricky to maintain.
Please use `URLSearchParams()` or something similar to create this in a more robust way.